be/src/fs/fs_s3.cpp

```diff
@@ -111,18 +118,63 @@ class S3ClientFactory {
 
 S3ClientFactory::S3ClientFactory() : _rand((int)::time(nullptr)) {}
 
+// Get a AWSCredentialsProvider based on CloudCredential
+std::shared_ptr<Aws::Auth::AWSCredentialsProvider> S3ClientFactory::_get_aws_credentials_provider(
+        const TAWSCloudCredential& aws_cloud_credential) {
+    // Do some check
+    DCHECK(aws_cloud_credential.__isset.use_aws_sdk_default_behavior)
+            << "use_aws_sdk_default_behavior value must set in FE";
+    DCHECK(aws_cloud_credential.__isset.use_instance_profile) << "use_instance_profile value must set in FE";
+
+    std::shared_ptr<Aws::Auth::AWSCredentialsProvider> credential_provider = nullptr;
+    if (aws_cloud_credential.use_aws_sdk_default_behavior) {
+        credential_provider = std::make_shared<Aws::Auth::DefaultAWSCredentialsProviderChain>();
+    } else {
+        // Create a base credentials provider
+        if (aws_cloud_credential.use_instance_profile) {
+            credential_provider = std::make_shared<Aws::Auth::InstanceProfileCredentialsProvider>();
+        } else if (aws_cloud_credential.__isset.access_key && aws_cloud_credential.__isset.secret_key) {
+            credential_provider = std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(
+                    aws_cloud_credential.access_key, aws_cloud_credential.secret_key);
+        } else {
+            DCHECK(false) << "Unreachable!";
+            credential_provider = std::make_shared<Aws::Auth::AnonymousAWSCredentialsProvider>();
+        }
+
+        if (aws_cloud_credential.__isset.iam_role_arn) {
+            // Do assume role
+            auto sts = std::make_shared<Aws::STS::STSClient>(credential_provider);
+            credential_provider = std::make_shared<Aws::Auth::STSAssumeRoleCredentialsProvider>(
+                    aws_cloud_credential.iam_role_arn, Aws::String(), aws_cloud_credential.external_id,
+                    Aws::Auth::DEFAULT_CREDS_LOAD_FREQ_SECONDS, sts);
+        }
+    }
+    return credential_provider;
+}
+
 S3ClientFactory::S3ClientPtr S3ClientFactory::new_client(const ClientConfiguration& config, const FSOptions& opts) {
     std::lock_guard l(_lock);
 
     for (size_t i = 0; i < _items; i++) {
         if (_configs[i] == config) return _clients[i];
     }
 
+    const THdfsProperties* hdfs_properties = opts.get_hdfs_properties();
```

Comment:
- this can be run of of lock, see LOC:156
- the early return will pass the cache logic
- maybe it's better to introduce lru cache logic for caching clients